main.go

```diff
@@ -153,3 +150,75 @@ func executeCheck(event *types.Event) (int, error) {
 
 	return sensu.CheckStateOK, nil
 }
+
+func getAllServices(client *http.Client) ([]string, error) {
+	req, err := http.NewRequest("GET", plugin.SupervisorURL+"/services", nil)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Accept", "application/json")
+
+	var resp *http.Response
+	resp, err = getResponse(client, req)
+	if err != nil {
+		return nil, err
+	}
+
+	var services ServiceResponse
+	if err := json.NewDecoder(resp.Body).Decode(&services); err != nil {
+		return nil, fmt.Errorf("failed to decode service response: %v", err)
+	}
+
+	var result = make([]string, len(services))
+	for i, v := range services {
+		result[i] = v.ServiceGroup
+	}
+
+	return result, nil
+}
+
+func checkServices(services []string, client *http.Client) ([]HealthResponse, error) {
+	var result []HealthResponse
+
+	for _, service := range services {
+		serviceSplit := strings.SplitN(service, ".", 2)
+
+		req, err := http.NewRequest("GET", plugin.SupervisorURL+"/services/"+serviceSplit[0]+"/"+serviceSplit[1]+"/health", nil)
+		if err != nil {
+			return nil, err
+		}
+
+		req.Header.Set("Accept", "application/json")
+
+		var resp *http.Response
+		resp, err = getResponse(client, req)
+		if err != nil {
+			return nil, err
+		}
+
+		var health HealthResponse
+		if err := json.NewDecoder(resp.Body).Decode(&health); err != nil {
+			return nil, fmt.Errorf("failed to decode health response: %v", err)
+		}
+		health.ServiceGroup = service
+
+		result = append(result, health)
+	}
+
+	return result, nil
+}
+
+func getResponse(client *http.Client, req *http.Request) (*http.Response, error) {
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
```

Comment:
In golang, defer delays execution of a function call until the enclosing function returns. In this case, we will close the body immediately due to this function returning and then try to access it in the calling function, which will result in an error.

We would need to move the body close out of this function to the call site, which limits the usefulness of the abstraction. We probably just want to drop this function and replace calls to getResponse with the direct Client.Do/Body.Close interaction.

Reply:
Interesting. I can see how `defer` might be useful. I will eliminate that utility function. It made more sense to have it before the second commit.